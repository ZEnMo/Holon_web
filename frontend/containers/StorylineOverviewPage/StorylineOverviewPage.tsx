import React from "react";
import StorylineOverview from "@/components/Storyline/StorylineOverview/StorylineOverview";
import styles from "./StorylineOverviewPage.module.css";

const StorylineOverviewPage = ({ allInformationTypes, allRoles, allStorylines }) => {
  console.log(allInformationTypes);
  return (
    <div className={styles["StorylineOverviewPage"]}>
      <StorylineOverview
        storylines={allStorylines}
        allInformationTypes={allInformationTypes}
        allRoles={allRoles}
      />
    </div>
  );
};

export default StorylineOverviewPage;
