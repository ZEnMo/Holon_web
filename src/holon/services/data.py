import json
from pathlib import Path
from typing import Any

import numpy as np

from holon.services.cloudclient.output import AnyLogicOutput
from pipit.settings import get_env_bool


class Results:
    def __init__(
        self,
        cc_payload: dict,
        request,
        anylogic_outcomes: AnyLogicOutput,
        inter_upscaling_outcomes: dict,
        nat_upscaling_outcomes: dict,
        cost_outcome: float,
        cost_benefit_overview: dict,
        cost_benefit_detail: dict,
        anylogic_outputs: dict[str, Any],
        datamodel_query_results: dict[int, Any],
        error: Exception = None,
    ) -> None:
        self.cc_payload = cc_payload
        self.request = request
        self.anylogic_outcomes = anylogic_outcomes
        self.inter_upscaling_outcomes = inter_upscaling_outcomes
        self.nat_upscaling_outcomes = nat_upscaling_outcomes
        self.cost_outcome = cost_outcome
        self.cost_benefit_overview = cost_benefit_overview
        self.cost_benefit_detail = cost_benefit_detail
        self.anylogic_outputs = anylogic_outputs
        self.datamodel_query_results = datamodel_query_results
        self.error = error

    @property
    def anylogic_outcomes(self):
        return self._anylogic_outcomes

    @anylogic_outcomes.setter
    def anylogic_outcomes(self, anylogic_outcomes: AnyLogicOutput):
        self._anylogic_outcomes = calculate_holon_kpis(anylogic_outcomes)

    def to_dict(self):
        result = {
            "dashboard_results": {
                "local": {**self.anylogic_outcomes, "costs": self.cost_outcome},
                "intermediate": self.inter_upscaling_outcomes,
                "national": self.nat_upscaling_outcomes,
            },
            "cost_benefit_results": {
                "overview": self.cost_benefit_overview,
                "detail": self.cost_benefit_detail,
            },
            "anylogic_outputs": self.anylogic_outputs,
            "datamodel_query_results": self.datamodel_query_results,
            "error": repr(self.error) if self.error is not None else None,
        }
        if self.__include_scenario():
            result["scenario"] = self.cc_payload

        return result

    def __include_scenario(self):
        """Only include modified scenario if env variable is set"""
        return get_env_bool("RETURN_SCENARIO", False)


def calculate_holon_kpis(anylogic_outcomes: AnyLogicOutput) -> dict:
    def get_key_over_all_results(key):
        return anylogic_outcomes.get_key_over_all_results(key)

    import_curve_MWh = np.array(
        list(get_key_over_all_results("SystemHourlyElectricityImport_MWh").values())[:8760]
    )

    jsons = Path(__file__).absolute().parent / "jsons"

    with open(jsons / "share_of_renewable_electricity.json", "r") as _infile:
        share_of_renewables = np.array(json.load(_infile))

    UnsustainableImportedElectricity_MWh = np.inner(1 - share_of_renewables, import_curve_MWh)

    Sustainability_pct = 100 * (
        1
        - (
            get_key_over_all_results("totalDieselImport_MWh")
            + get_key_over_all_results("totalMethaneImport_MWh")
            # assumption is that all imported hydrogen is grey
            + get_key_over_all_results("totalHydrogenImport_MWh")
            + UnsustainableImportedElectricity_MWh
        )
        / get_key_over_all_results("TotalEnergyUsed_MWh")
    )
    Sustainability_pct = min(100, Sustainability_pct)

    netOverload_pct = get_key_over_all_results("netOverload_pct")
    if netOverload_pct == 1:
        netload_mv_pos_pct = (
            (get_key_over_all_results("MSLSnodePeakPositiveLoadElectricity_kW"))
            / get_key_over_all_results("cumulativeCapacityLS")
        ) * 100

        netload_mv_neg_pct = (
            (get_key_over_all_results("MSLSnodePeakNegativeLoadElectricity_kW"))
            / get_key_over_all_results("cumulativeCapacityLS")
        ) * 100

        netOverload_pct = max(abs(netload_mv_pos_pct), abs(netload_mv_neg_pct))

    KPIs = {
        "sustainability": round(Sustainability_pct, 1),
        "self_sufficiency": round(get_key_over_all_results("totalSelfSufficiency_fr") * 100, 1),
        "netload": round(netOverload_pct, 1),
    }
    return KPIs
