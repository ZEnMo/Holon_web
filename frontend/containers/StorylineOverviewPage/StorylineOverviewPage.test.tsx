import { render /* screen */ } from "@testing-library/react";
import StorylineOverviewPage from "./";
import data from "./StorylineOverviewPage.data";

describe("<StorylineOverviewPage />", () => {
  it("Renders an empty StorylineOverviewPage", () => {
    render(<StorylineOverviewPage />);
  });

  it("Renders StorylineOverviewPage with data", () => {
    const { container } = render(<StorylineOverviewPage {...data} />);
    expect(container).toMatchSnapshot();
  });
});
