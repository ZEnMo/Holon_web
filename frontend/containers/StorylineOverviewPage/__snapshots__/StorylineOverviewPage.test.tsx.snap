// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StorylineOverviewPage /> Renders StorylineOverviewPage with data 1`] = `
<div>
  <div
    class="StorylineOverviewPage"
  >
    <div
      class="flex w-full flex-col lg:flex-row"
    >
      <div
        class="flex flex-col p-8 lg:w-1/3 xl:w-1/4"
      >
        <div
          class="flex w-full flex-col gap-4 mb-5"
        >
          <h4
            class="relative my-4 border-l-[0.75rem] border-b-2 border-holon-blue-900 pl-3 text-xl font-light"
          >
            Type rol
          </h4>
        </div>
        <div
          class="flex w-full flex-col gap-4 mb-5"
        >
          <h4
            class="relative my-4 border-l-[0.75rem] border-b-2 border-holon-blue-900 pl-3 text-xl font-light"
          >
            Type informatie
          </h4>
        </div>
      </div>
      <div
        class="flex flex-col p-8 lg:w-2/3 xl:w-3/4"
      >
        <div
          class="flex flex-row justify-between mb-2"
        >
          <strong>
             resultaten
          </strong>
          <strong>
            Sorteren placenholder
          </strong>
        </div>
        <div
          class="flex flex-row flex-wrap storyline__grid"
        />
      </div>
    </div>
  </div>
</div>
`;
